async: Do not use heapless on public API.

Dirbaio opened the pull request:
It's better to not use heapless on the public API.

- It can get major bumps from time to time (for example v0.8 was just relased), and upgrading it is a breaking change of embedded-nal-async, which we should try to avoid.
- It allows users to use any version of heapless, or something else entirely.

Dirbaio committed:
async: fix AFIT warnings on latest nightlies.

lulf reviewed (approved):
Fully agreed! This bothered me with the existing blocking api, should that be changed too?

ryan-summers commented:
Would you be willing to copy these over to the non-async version as well? I agree with the sentiment here, and we have the same problem over there.

Dirbaio committed:
Switch to Edition 2021, bump MSRV.

Dirbaio committed:
Do not use heapless on public API.

It's better to not use heapless on the public API.

- It can get major bumps from time to time (for example v0.8 was just relased), and upgrading it is a breaking change of embedded-nal-async, which we should try to avoid.
- It allows users to use any version of heapless, or something else entirely.

Dirbaio commented:
@ryan-summers  done

also note I've separated the MSRV bump to #94, review that first.

ryan-summers reviewed (approved):
Thanks for fixing this!
